modules/sdk/gradle-plugins/src/main/java/com/liferay/gradle/plugins/ModuleExtPlugin.java

```diff
@@ -0,0 +1,292 @@
+
+import java.io.File;
+import java.io.FileInputStream;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.jar.Attributes;
+import java.util.jar.Attributes.Name;
+
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+import org.gradle.api.Task;
+import org.gradle.api.artifacts.Configuration;
+import org.gradle.api.file.FileTree;
+import org.gradle.api.java.archives.Manifest;
+import org.gradle.api.plugins.JavaPlugin;
+import org.gradle.api.tasks.Copy;
+import org.gradle.api.tasks.SourceSet;
+import org.gradle.api.tasks.SourceSetOutput;
+import org.gradle.api.tasks.bundling.Jar;
+
+import org.osgi.framework.Version;
+
+/**
+ * @author David Truong
+ */
+public class ModuleExtPlugin implements Plugin<Project> {
+
+	public static final String MODULE_EXT_CONFIGURATION_NAME = "moduleExt";
+
+	public static final String UNZIP_MODULE_EXT_JAVA_TASK_NAME =
+		"unzipModuleExtJava";
+
+	public static final String UNZIP_MODULE_EXT_MANIFEST_TASK_NAME =
+		"unzipModuleExtManifest";
+
+	public static final String UNZIP_MODULE_EXT_RESOURCES_TASK_NAME =
+		"unzipModuleExtResources";
+
+	@Override
+	public void apply(final Project project) {
+		GradleUtil.applyPlugin(project, JavaPlugin.class);
+		GradleUtil.applyPlugin(project, LiferayBasePlugin.class);
+
+		Configuration moduleExtConfiguration = _addConfigurationModuleExt(
+			project);
+
+		_addTaskUnzipModulesExtJava(project, moduleExtConfiguration);
+
+		_addTaskUnzipModulesExtResources(project, moduleExtConfiguration);
+
+		Copy copy = _addTaskUnzipModulesExtManifest(
+			project, moduleExtConfiguration);
+
+		Jar jar = _configureJarTask(project, copy);
+
+		_configureDeployTask(project, jar);
+	}
+
+	private Configuration _addConfigurationModuleExt(final Project project) {
+		Configuration configuration = GradleUtil.addConfiguration(
+			project, MODULE_EXT_CONFIGURATION_NAME);
+
+		configuration.setDescription(
+			"Configures the module you would like to extend.");
+		configuration.setTransitive(true);
+		configuration.setVisible(false);
+
+		Configuration compileOnlyConfiguration = GradleUtil.getConfiguration(
+			project, JavaPlugin.COMPILE_ONLY_CONFIGURATION_NAME);
+
+		compileOnlyConfiguration.extendsFrom(configuration);
+
+		return configuration;
+	}
+
+	private Copy _addTaskUnzipModulesExtJava(
+		final Project project, final Configuration configuration) {
+
+		Copy copy = GradleUtil.addTask(
+			project, UNZIP_MODULE_EXT_JAVA_TASK_NAME, Copy.class);
+
+		copy.from(
+			new Callable<FileTree>() {
+
+				@Override
+				public FileTree call() {
+					return project.zipTree(configuration.getSingleFile());
+				}
+
+			});
+
+		copy.include("**/*.class");
+		copy.setIncludeEmptyDirs(false);
+
+		SourceSet sourceSet = GradleUtil.getSourceSet(
+			project, SourceSet.MAIN_SOURCE_SET_NAME);
+
+		SourceSetOutput sourceSetOutput = sourceSet.getOutput();
+
+		copy.setDestinationDir(sourceSetOutput.getClassesDir());
```

Comment:
What if someone changes `sourceSets.main.output.classesDir` after applying this plugin? Shouldn't we unzip in a temporary directory, and simply configure the `jar` task to copy `from unzipModulesExt`? Then maybe you need one single task instead of 3.

Reply:
I need to ensure that the files get copied first into the jar first before we copy the newly compiled classes and any new resources... could we do that from a doFirst?